influunt-app/app/scripts/services/alarmes-dinamico-service.js

```diff
@@ -0,0 +1,359 @@
+        var statusObj, _fnOnEventTriggered, controladores;
+        var _fnonClickToast = function(){};
+
+        var registerWatchers = function() {
+          pahoProvider.connect()
+            .then(function() {
+              pahoProvider.register(eventosDinamicos.STATUS_CONTROLADORES, statusControladoresWatcher);
+              pahoProvider.register(eventosDinamicos.ALARMES_FALHAS, alarmesEFalhasWatcher);
+              pahoProvider.register(eventosDinamicos.TROCA_PLANO, trocaPlanoWatcher);
+              pahoProvider.register(eventosDinamicos.CONTROLADOR_ONLINE, onlineOfflineWatcher);
+              pahoProvider.register(eventosDinamicos.CONTROLADOR_OFFLINE, onlineOfflineWatcher);
+            });
+        };
+
+        var unregisterWatchers = function() {
+          pahoProvider.connect()
+            .then(function() {
+              pahoProvider.unregister(eventosDinamicos.STATUS_CONTROLADORES);
+              pahoProvider.unregister(eventosDinamicos.ALARMES_FALHAS);
+              pahoProvider.unregister(eventosDinamicos.TROCA_PLANO);
+              pahoProvider.unregister(eventosDinamicos.CONTROLADOR_ONLINE);
+              pahoProvider.unregister(eventosDinamicos.CONTROLADOR_OFFLINE);
+            });
+        };
+
+
+        // watchers.
+        statusControladoresWatcher = function(payload) {
+          var mensagem = JSON.parse(payload);
+          mensagem.conteudo = _.isString(mensagem.conteudo) ? JSON.parse(mensagem.conteudo) : mensagem.conteudo;
+          statusObj.status = statusObj.status || {};
+
+          return getControlador(mensagem.idControlador).then(function(controlador) {
+            if (!controlador) {
+              console.log('controlador', mensagem.idControlador, 'não existe.');
+              return false;
+            }
+
+            controlador.statusControlador = mensagem.conteudo.status;
+            controlador.status = mensagem.conteudo.status;
+            statusObj.status[mensagem.idControlador] = _.get(mensagem, 'conteudo.status');
+
+            if (isAlertaAtivado(mensagem.tipoMensagem)) {
+              var msg = $filter('translate')(
+                'controladores.mapaControladores.alertas.mudancaStatusControlador',
+                {CONTROLADOR: controlador.CLC}
+              );
+
+              exibirAlerta(msg, controlador);
+            }
+
+            return _.isFunction(_fnOnEventTriggered) && _fnOnEventTriggered.apply(this, mensagem);
+          });
+        };
+
+        alarmesEFalhasWatcher = function(payload) {
+          var mensagem = JSON.parse(payload);
+          mensagem.conteudo = _.isString(mensagem.conteudo) ? JSON.parse(mensagem.conteudo) : mensagem.conteudo;
+          statusObj.erros = statusObj.erros || {};
+
+          getControlador(mensagem.idControlador).then(function(controlador) {
+            if (!controlador) {
+              console.log('controlador', mensagem.idControlador, 'não existe.');
+              return false;
+            }
```

Comment:
Esse primeiro trecho do `alarmesEFalhasWatcher` é idêntico ao início dos outros watchers, é possível extrair para uma função? Talvez para dentro do `getControlador()` e aí a promise só resolver se achar o controlador, por exemplo.